### PR TITLE
[refactor] Refactor result buffer to allow return values and error codes coexist

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #886 #1016 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/inc/constants.h [hunk 1]
```diff
@@ -10,6 +10,11 @@ constexpr std::size_t taichi_global_tmp_buffer_size = 1024 * 1024;
 constexpr int taichi_max_num_mem_requests = 1024 * 64;
 constexpr std::size_t taichi_page_size = 4096;
 constexpr std::size_t taichi_max_message_length = 2048;
+constexpr std::size_t taichi_result_buffer_entries = 32;
```
Comment (xumingkuan) [inline taichi/inc/constants.h:13]: Does this mean that we can return 32 `uint64`s? But it seems that only the entries `0` and `1` are used?

File: taichi/program/program.cpp [hunk 2]
```diff
@@ -310,11 +313,11 @@ void Program::check_runtime_error() {
...
-  auto error_code = runtime_jit_module->fetch_result<int64>();
+  auto error_code = fetch_result<int64>(taichi_result_buffer_error_id);
   if (error_code) {
     runtime_jit_module->call<void *>("runtime_retrieve_error_message",
                                      llvm_runtime);
-    auto error_message = runtime_jit_module->fetch_result<char *>();
+    auto error_message = fetch_result<char *>(taichi_result_buffer_error_id);
     if (error_code == 1) {
       TI_ERROR("Assertion failure: {}", error_message);
     } else {
```
Comment (archibate) [inline taichi/program/program.cpp:323]: So, it's actually:
```
union {
  int64 error_code;
  char *error_message;
};
```
?
They why not just `if (error_message != nullptr)`?

Comment (xumingkuan) [reply taichi/program/program.cpp:323]: `runtime_retrieve_error_message` should set the pointer. It's not a `union`.

File: taichi/inc/constants.h (hunk 1 above)
Comment (yuanming-hu) [reply taichi/inc/constants.h:13]: We are using a buffer with `32` entries. Unused buffer entries are not fetched back.

File: taichi/runtime/llvm/runtime.cpp [hunk 3]
```diff
@@ -525,8 +525,8 @@ struct LLVMRuntime {
   i32 allocator_lock;
 
   template <typename T>
-  void set_result(T t) {
-    *(u64 *)result_buffer = taichi_union_cast<uint64>(t);
+  void set_result(std::size_t i, T t) {
+    ((u64 *)result_buffer)[i] = taichi_union_cast<uint64>(t);
```
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply taichi/runtime/llvm/runtime.cpp:529]: They actually have the same value but different address spaces. The pointer in `LLVMRuntime` is available to the compiled kernels.

File: taichi/inc/constants.h [hunk 4]
```diff
@@ -10,6 +10,11 @@ constexpr std::size_t taichi_global_tmp_buffer_size = 1024 * 1024;
...
+// slot for kernel return value
+constexpr std::size_t taichi_result_buffer_ret_value_id = 0;
+// slot for error code and error message char *
+constexpr std::size_t taichi_result_buffer_error_id = 1;
```
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply taichi/inc/constants.h:17]: Yeah, I thought about that. But in this PR let's stick to what we have now.

**Pull request merged**